jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,735 @@
+        {
+            if (p instanceof MethodPredicate methodPredicate)
+                excludeMethod(methodPredicate._method);
+            else if (p instanceof PathSpecPredicate pathSpecPredicate)
+                exclude(pathSpecPredicate._pathSpec);
+            else
+                _predicates.exclude(p);
+        }
+    }
+
+    private boolean testMethods(Request request)
+    {
+        return _methods.test(request.getMethod());
+    }
+
+    private boolean testPathSpecs(Request request)
+    {
+        return _pathSpecs.test(Request.getPathInContext(request));
+    }
+
+    private boolean testPredicates(Request request)
+    {
+        return _predicates.test(request);
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _handlePredicate = TypeUtil.truePredicate();
+
+        if (!_methods.isEmpty())
+            _handlePredicate = _handlePredicate.and(this::testMethods);
+        if (!_pathSpecs.isEmpty())
+            _handlePredicate = _handlePredicate.and(this::testPathSpecs);
+        if (!_predicates.isEmpty())
+            _handlePredicate = _handlePredicate.and(this::testPredicates);
+
+        super.doStart();
+    }
+
+    public final boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        // This is the code we would have in handle if MethodHandlers were not used
+        if (_handlePredicate.test(request))
+            return doHandle(request, response, callback);
+
+        return doNotHandle(request, response, callback);
+    }
+
+    /**
+     * Handle a request that has met the conditions.
+     * Typically, the implementation will provide optional handling and then call the
+     * {@link #nextHandle(Request, Response, Callback)} method to continue handling.
+     * @param request The request to handle
+     * @param response The response to generate
+     * @param callback The callback for completion
+     * @return True if this handler will complete the callback
+     * @throws Exception If there is a problem handling
+     * @see Handler#handle(Request, Response, Callback)
+     */
+    protected abstract boolean doHandle(Request request, Response response, Callback callback) throws Exception;
+
+    /**
+     * Handle a request by invoking the {@link #handle(Request, Response, Callback)} method of the
+     * {@link #getHandler() next Handler}.
+     * @param request The request to handle
+     * @param response The response to generate
+     * @param callback The callback for completion
+     * @return True if this handler will complete the callback
+     * @throws Exception If there is a problem handling
+     * @see Handler#handle(Request, Response, Callback)
+     */
+    protected boolean nextHandle(Request request, Response response, Callback callback) throws Exception
+    {
+        Handler next = getHandler();
+        return next != null && next.handle(request, response, callback);
```

Comment:
This should be `return super.handle(request, response, callback)`.

Reply:
yeah... but I kind of like taking the stack frame out with the manual inline here.  I no it probably saves nothing of significance, but we have often done such inlining before.  When looking at a stack,  it is nice to see the `nextHandle` method directly invoke the next handler method, rather than see a `Handler.Wrapper#handle` in between.

I've removed the inline under protest.